packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,165 @@ class UpgradeCommand extends FlutterCommand {
-      <String>['git', 'pull', '--ff-only'],
+  /// Attempts to rebase the upstream onto the local branch.
+  ///
+  /// If there haven't been any hot fixes or local changes, this is equivalent
+  /// to a fast-forward.
+  Future<void> attemptRebase() async {
+    final int code = await runCommandAndStreamOutput(
+      <String>['git', 'pull', '--rebase'],
       workingDirectory: Cache.flutterRoot,
       mapFunction: (String line) => matchesGitLine(line) ? null : line,
     );
-
-    if (code != 0)
+    if (code != 0) {
       throwToolExit(null, exitCode: code);
+    }
+  }
 
-    // Check for and download any engine and pkg/ updates.
-    // We run the 'flutter' shell script re-entrantly here
-    // so that it will download the updated Dart and so forth
-    // if necessary.
+  /// Update the engine repository and precache all artifacts.
+  ///
+  /// Check for and download any engine and pkg/ updates. We run the 'flutter'
+  /// shell script re-entrantly here so that it will download the updated
+  /// Dart and so forth if necessary.
+  Future<void> precacheArtifacts() async {
     printStatus('');
     printStatus('Upgrading engine...');
-    code = await runCommandAndStreamOutput(
+    final int code = await runCommandAndStreamOutput(
       <String>[
         fs.path.join('bin', 'flutter'), '--no-color', '--no-version-check', 'precache',
       ],
       workingDirectory: Cache.flutterRoot,
       allowReentrantFlutter: true,
     );
+    if (code != 0) {
+      throwToolExit(null, exitCode: code);
+    }
+  }
 
+  /// Update the user's packages.
+  Future<void> updatePackages(FlutterVersion flutterVersion) async {
     printStatus('');
     printStatus(flutterVersion.toString());
-
     final String projectRoot = findProjectRoot();
     if (projectRoot != null) {
       printStatus('');
       await pubGet(context: PubContext.pubUpgrade, directory: projectRoot, upgrade: true, checkLastModified: false);
     }
+  }
 
-    // Run a doctor check in case system requirements have changed.
+  /// Run flutter doctor in case requirements have changed.
+  Future<void> runDoctor() async {
     printStatus('');
     printStatus('Running flutter doctor...');
-    code = await runCommandAndStreamOutput(
+    await runCommandAndStreamOutput(
       <String>[
         fs.path.join('bin', 'flutter'), '--no-version-check', 'doctor',
       ],
       workingDirectory: Cache.flutterRoot,
       allowReentrantFlutter: true,
     );
+  }
 
-    return null;
+  /// Pop stash changes if [stashName] is non-null and contained in stash.
+  Future<void> applyStash(String stashName) async {
+    if (stashName == null) {
+      return;
+    }
+    try {
+      final RunResult result = await runCheckedAsync(<String>[
+        'git', 'stash', 'list'
+      ]);
+      if (!result.stdout.contains(stashName)) {
+        // print the same warning as if this threw.
+        throw Exception();
+      }
+      await runCheckedAsync(<String>[
+        'git', 'stash', 'pop',
```

Comment:
`apply` instead of `pop` (and I think you'll have to pass the stash name)

Reply:
That would be the stash index though, right? IIRC you can't name a stash, only leave a comment with it

Reply:
You're right.  You'd have to call `git stash list` and grep for the topmost match of `stashName` to find out the id (index) of the stash

Reply:
If we succeeded our stash will always be the topmost, which we confirm with the list check above

Reply:
Good point.  And if the pop fails for any reason, they'll still be left with the stash.

Reply:
why do we check the list if we know it'll be topmost?